### PR TITLE
execution/stagedsync: compute a tx's fee credit once, not on every revalidation

### PR DESCRIPTION
**Pull request by AskAlexSharov**
Parallel execution credits the miner at the end of each tx, and redoes that work every time the tx is revalidated — 4-5 times per tx at the chain tip, up to ~30 while catching up. The credit changes only when an earlier tx moved the coinbase balance, so it is now computed once: a later round compares it with what the tx already carries and does nothing when they match.

One bug fixed on the way, already on `main` and not caused by the skip. A tx with a zero tip on an empty coinbase records a delete of that coinbase (EIP-161). If an earlier tx then funds the coinbase, this tx has no fee adjustment left to make, but its delete stayed recorded and stayed in the version map, so a later tx could read the coinbase as deleted. The round now takes it back out. `release/3.6` has the same bug (it is in merged #23131); the backport follows.

**Not merge-ready**: needs an n5 soak with `grep -ci "Wrong trie root"`. Green unit suites are not evidence here — see the first block below.

`BenchmarkCalcFees`, 200000x, n=4, laptop. The benchmark is new, so there is no `main` column: `main` runs the `first credit` path on every round, this PR runs it once per tx and every round after it costs the `repeat round` column.

| BenchmarkCalcFees | first credit | repeat round |
|---|---|---|
| pre_london | 204 ns, 6 allocs, 528 B | 110 ns, 2 allocs, 192 B |
| london | 325 ns, 11 allocs, 976 B | 209 ns, 4 allocs, 384 B |

The skipped round also drops the `MergeInto` over the tx's writes and the map release after it, which these numbers do not cover.

<details>
<summary>Why the credit is compared against the fee-merge product and not against the tx's own writes</summary>

`calcFees` reads the tx's own write set (`TxOut`) as the pre-credit balance, so folding the credit into it makes every round add the tip again. That is what `WriteSet.Absorb` did on `alex/fee_merge_in_place_37`: 715 `Wrong trie root` errors on n5 within a minute, with unit suites green the whole time. The comparison here is against `feeMergeTemp[tx]` instead — a separate set `calcFees` never reads, dropped as soon as the tx re-executes.
</details>

<details>
<summary>Why the outcome is three-way</summary>

A single nil return cannot tell "already recorded" from "nothing to emit", and those two need opposite handling: keep the recorded credit, or take it back out. `calcFees` now reports `feeCreditRecorded` / `feeCreditNew` / `feeCreditNone`, and the fee merge keeps the worker's own `TxOut` as the base it merges onto — so an entry an earlier credit emitted and this one does not cannot survive the merge either. Headers the new set drops are deleted from the version map, the same way the re-execution path does it.
</details>

<details>
<summary>What the skip does not cover</summary>

When the coinbase is also the burnt-fee address and both halves take the balance arm, they can never match at once — one write cannot carry both tracing reasons — so such blocks pay the full rebuild every round.
</details>

<details>
<summary>Difference from #23131 (release/3.6)</summary>

Same change plus the retraction above. `main` has no `CollectorWrites` on the result, so the two `SetAccountBalanceOrDelete` calls that sit next to the emission there are absent here, along with the skip's ordering ahead of them and the matching test-harness line. Nothing else differs.
</details>

### CODE REVIEW

**Comment by AskAlexSharov** [comment]
Carries the review follow-ups from #23131 (erigontech/erigon#23131 (review)): `recordWorkerWrites` clearing `feeMergeTemp` on re-execution, lazy `feeEntry` construction, and the EIP-161 delete + `recordedIn`/`writeTo` round-trip tests.

The `CollectorWrites` half of that feedback does not apply here — `main` has no such field, so there was never anything above the check to move.

**Review by yperbasis — CHANGES_REQUESTED** [review]
Combined review — my pass plus the Codex finding folded in. No correctness bug found: every skip mismatch fails toward a full re-credit. Remaining items by severity.

**High**

- `recordWorkerWrites` (exec3_parallel.go:2500) deletes the displaced fee-merge product from `feeMergeTemp` without `queueMapRelease`, so its maps go to GC on every re-execution of an already-credited tx (same as the Codex point). Not a regression — base leaked the same way — but the new helper is the natural owner of the release. Commit d85b484fe7d51a6386f5f915a89ee82cdb998f95 on the 3.6 twin (`alex/calcfees_skip_recredit_36`) already implements the model: release in `recordWorkerWrites`, version-pinned `feeMergeTemp`, `recordFeeMerge` owning the merge+record, and `TestRecordWorkerWritesDropsCreditedTemp` asserting the release. Please port it here; it also closes the first medium item and the first low item.

**Medium**

- The finalize-path `RecordWrites` (exec3_parallel.go:2957) bypasses the `feeMergeTemp` protocol. Latent today because `finalizeTx` returns nil writes for regular txs, but if that changes, a stale temp silently disables the skip and its maps become unreclaimable. The version pin in the ported commit closes this.
- `feeEntry`'s zero value reads as "already recorded" (exec3_parallel.go:2092): a future construction site that forgets `emit: true` makes `recordedIn` vacuously true and `writeTo` a no-op, so the credit is dropped with no error and no failing test. Nil `*feeEntry` = absent would make that state unrepresentable.
- `recordedIn` compares `*aw.Val == e.acc` (exec3_parallel.go:2109), stricter than `Account.Equals` because it also includes Root and PrevIncarnation. If `feeAddressAccount` ever fills those asymmetrically, the skip silently stops firing — an invisible perf regression. Use an `Equals`-based compare, or add a comment plus a symmetry test row.
- Test gap (exec3_fee_credit_test.go:153): the deleted arm of `recordedIn` accepts any same-version SelfDestruct write, more than the docstring claims; the fences that make this benign (worker writes stamped with TxNum=0) are not pinned by any test. Add a foreign-SD rejection row or soften the docstring.

**Low**

- The temp-identity predicate exists twice: `creditedWrites` (exec3_parallel.go:2509) and `recordFeeMerge` (line 2521). The ported commit dissolves it.
- `feeEntry.acc` and `.reason` are dead in the deleted state (exec3_parallel.go:2056); populate them only in the credit arm.
- When the coinbase is the burnt contract address, the skip can never fire: one recorded write cannot carry both tracing reasons (exec3_parallel.go:2073). Perf-only and pathological; a short comment is enough.
- `newFeeCreditRound` (exec3_fee_credit_test.go:30) re-copies the `runFinalizeTx` scaffold, and `recorded`/`credited` alias the same pointer after the first round (line 61). The ported commit's test consolidation is a good moment to fix both.

Follow-up outside this PR: with the skip keeping the set identity stable, the remaining per-round cost is the unconditional `FlushVersionedWrites` of an unchanged set (exec3_parallel.go:2869); a (set pointer, flag) memo could remove it.

**Comment by AskAlexSharov** [comment]
All items addressed on current head. Package suite green; `BenchmarkCalcFees` unchanged — 195 ns/op, 6 allocs for the first credit against 87 ns/op, 2 allocs for the skipped round, so the skip still fires.

**High**

Ported. `recordWorkerWrites` takes the version instead of the index, `queueMapRelease`s the displaced product before dropping it, and `recordFeeMerge` now owns the `MergeInto` and the `RecordWrites` — so the merge product and the tx's recorded set cannot drift apart, and the "a temp is never an `execResult`'s TxOut" rule is local to the one function that populates the map.

I drove the leak red before porting: with the old `recordWorkerWrites`, `tip.Count()` after a re-execution is 1, not 0. `TestRecordWorkerWritesDropsCreditedTemp` is that assertion.

**Medium**

1. *Finalize-path `RecordWrites`* — closed by the version pin, as you said. `feeMergeTemp` now stores `feeMerge{writes, version}` and `creditedWrites` rejects anything whose version is not the caller's, so a temp that outlives its recorded set cannot be handed back. `TestCreditedWritesPinsVersion` covers the re-executed incarnation, another tx's product, and a nil set.

2. *`feeEntry`'s zero value* — took your suggestion: absent is `nil *feeEntry`, `emit` is gone. `recordedIn` and `writeTo` are nil-receiver methods, so a construction site that skips an entry gets the absent behaviour by not building it, and there is no longer a field to forget. `TestFeeEntryNilIsAbsent` pins both halves.

3. *`recordedIn` stricter than `Equals`* — switched to `aw.Val.Equals(&e.acc)`. `Equals` compares exactly the four fields `feeAddressAccount` fills; `Root` and `PrevIncarnation` are set by neither side, so folding them in could only ever turn the skip off silently.

4. *Test gap on the deleted arm* — added `TestFeeEntryDeletedRejectsForeignSelfDestruct`, which pins the fence rather than softening the docstring: a `SelfDestruct` write left unstamped (the worker's own) is rejected, and so is a same-version write whose `Val` is false. The docstring now says the version stamp is what makes the arm safe.

**Low**

1. *Duplicated temp-identity predicate* — gone; `recordFeeMerge` is the only place that compares against `feeMergeTemp`.

2. *Dead `acc`/`reason` in the deleted state* — the coinbase entry fills them only in the credit arm now.

3. *Coinbase == burnt address* — comment added at the skip.

4. *`newFeeCreditRound`* — `run` returns a copy of the credit, so it stays observable after `MergeInto` folds the recorded set into it in place. The scaffold itself stays: unlike the 3.6 twin, `buildExecResult` on main does not populate `TxIn`/`TxOut`, so those two lines are doing real work. For the "not only the hand-mirrored fold" half I added `TestCalcFeesRoundThroughBlockExecutor`, which runs the round the way the apply loop does — `creditedWrites` in, `recordFeeMerge` out — against the real `feeMergeTemp` bookkeeping. `TestCreditedWrites` and `TestCreditedWritesAfterReExecution` are superseded by the new pair and removed.

The follow-up on the unconditional `FlushVersionedWrites` of an unchanged set is worth its own PR — the set identity is stable now, so the `(set pointer, flag)` memo you describe has something to key on. Not in this one.

**Review by sudeepdino008 — APPROVED** [review]

**Review by yperbasis — CHANGES_REQUESTED** [review]
### Blocking

1. `exec3_parallel.go:2050-2052` — the comment "A coinbase that is also the burnt address never matches: one write cannot carry both tracing reasons" is wrong. The deleted arm of `recordedIn` has no Reason check, so with coinbase == burnt address, an EIP-161-emptied coinbase and `FeeTipped == 0`, both `recordedIn` calls pass and the skip does fire. Harmless today (the rebuilt set would be identical), but please reword it to cover only the non-deleted shape.

2. `exec3_parallel.go:2076-2078` — the docstring case "a worker's SELFDESTRUCT carries the same version and reads as recorded" cannot happen in production: worker writes are stamped by `IntraBlockState.Version()` with `TxNum == 0`, while `writeTo` stamps the task version with `TxNum > 0`, so the full-struct Version compare rejects every worker write. That TxNum difference is the real fence on the deleted arm, and nothing documents or pins it — `TestFeeEntryDeletedMatchesAnyWriterAtThisVersion` builds both sides with `TxNum == 0`, a combination production can't produce. Please fix the docstring and the test premise (or compare a deliberate subset of Version fields).

3. `exec3_parallel.go:2079` — `recordedIn` is only tested in the accepting direction (`TestFeeEntryWriteToIsRecordedIn`); nothing asserts it rejects a set that differs in a component `writeTo` emits, and the rejecting checks are uneven (no Reason check on the SelfDestruct and AddressPath arms; `Account.Equals` ignores some fields). Please add a mutation-style test: build the recorded entry, flip each emitted field in turn, assert the skip is refused. The rejecting direction is the one that prevents a wrong skip — the failure class this path has already produced.

4. `exec3_parallel.go:2523` and `:2495` — both release sites lost the old self-release identity guard: `recordFeeMerge` releases `temp.writes` whenever it equals `prev` without checking `merged != temp.writes`, and `recordWorkerWrites` releases the displaced temp without checking it differs from the incoming set. Safety now rests on two unasserted facts (calcFees always returns a fresh set; a worker TxOut is never the recorded merge product); if a future change breaks either, we record a set and then zero it on the background release goroutine — silent write loss. A one-line identity check (or a `dbg` assert) restores the old defense cheaply.

### Non-blocking

- `exec3_parallel.go:2722-2746` — `prevWrites` aliases the set that `recordWorkerWrites` queues for background release; its uses are safe only because they come before that call. A short comment (or moving the record to the end of the arm) would keep a future edit from touching `prevWrites` after the release is queued.
- `exec3_parallel.go:2955` and `:3237` — these sites still record via bare `blockIO.RecordWrites`, bypassing the new `feeMergeTemp` bookkeeping. Fine while finalize returns nil writes for regular txs; an assert (or routing through a helper) would keep the invariant honest.
- `exec3_parallel.go:2821-2827` — `existingWrites` is read before `calcFees` and reused as `prev`. Fetching the recorded set inside `creditedWrites`/`recordFeeMerge` would remove the staleness gap and the parameter that exists only to pass it through.
- `exec3_fee_merge_temp_test.go:79` — the third transition's comment still attributes the "temp present, prev is a fresh TxOut" state to re-execution, but `recordWorkerWrites` now deletes the temp on re-execution; the finalize-merge path is the only remaining route to that state.
- `exec3_finalize_test.go:1823` — `BenchmarkCalcFees` only measures the pre-London scenario, so the burnt-address half of a mainnet round (second versioned read + second `recordedIn`) is invisible. `londonTransferScenario` already exists; worth a table row.
- The skip path still heap-allocates the boxed `ReadAccountData` returns in `calcFees` (~2 allocs per redundant round). A by-value read on the locally-built reader would make skip rounds allocation-free. Fine as a follow-up.
- Test dedup: `feeMergeTestAddr` duplicates `fAddr` (this PR uses both), and `feeCreditRound.run` re-implements the round protocol that `recordFeeMerge`/`creditedWrites` now own. Backing the fixture with a `*blockExecutor` would collapse the stand-in and the bridge test.
- `exec3_parallel.go:2488-2492` — the release-ownership rationale ("every other recorded set is some execResult's TxOut…") now appears in three comments; keep it once at the `feeMergeTemp` field and point there.

### Pre-existing, out of scope (also on main; deserve tracking issues)

- A merged fee credit is never retracted: when a later validation round computes no adjustment, `calcFees` returns at `exec3_parallel.go:2029` before the skip/merge logic, and its coinbase read is invisible to validation (empty ReadSet), so a revalidated-but-not-re-executed tx can carry a stale coinbase delete/credit into commitment.
- coinbase == burnt-contract address (reachable on Gnosis-family chains): `burntEntry.writeTo` overwrites the coinbase entry on the same address, dropping `FeeTipped`, while serial applies both credits — parallel/serial divergence.

**Review by yperbasis — APPROVED** [review]

**Review by yperbasis — CHANGES_REQUESTED** [review]
Blocking: the concern in [this thread](https://github.com/erigontech/erigon/pull/23132#discussion_r3774241761) is valid.

calcFees returns nil for both:

1. The recorded set already carries the exact credit.
2. The current round should emit no fee adjustment.

recordFeeMerge treats both outcomes as a no-op. If one round records an EIP-161 coinbase delete and an earlier transaction later makes that coinbase non-empty, revalidation preserves the stale delete. The fee reads are not in TxIn, so the worker result can validate without re-execution.

Please distinguish these outcomes, retract or replace the previous fee-merge product when the adjustment disappears, and add a regression test for the empty-removal to non-empty transition. This behavior predates both PRs and is also present in merged #23131, so the fix should be backported to release/3.6.

File: execution/stagedsync/exec3_parallel.go [hunk 1]
```diff
@@ -2454,16 +2479,61 @@ func (be *blockExecutor) invalidBlockResult(err error) *blockResult {
...
+// released: prev is otherwise some execResult's TxOut, which stays live.
 // MergeInto shares VersionedWrite pointers rather than the maps holding them,
-// so pooling prev's maps leaves the writes merged now holds intact.
-func (be *blockExecutor) recordFeeMerge(tx int, prev, merged *state.WriteSet) {
-	if temp := be.feeMergeTemp[tx]; temp != nil && temp == prev && merged != temp {
-		be.queueMapRelease(temp)
+// so pooling those maps leaves the merged writes intact.
+func (be *blockExecutor) recordFeeMerge(txVersion state.Version, prev, tipWrites *state.WriteSet) {
+	if tipWrites.IsEmpty() {
+		return
```
[the remark this answers is not in the document]
Comment (AskAlexSharov) [reply execution/stagedsync/exec3_parallel.go]: The transition is real, and I can name it precisely — but it is not introduced here, so I want to agree on scope before changing it.

**The case.** Coinbase `C`, zero-tip tx `K`, no burnt. Round 1: `C` is empty, the tip touch trips EIP-161, `coinbaseEmptied` is true, and `{C: SelfDestructPath}` is merged and recorded. Round 2, after an earlier tx `J` gives `C` a balance: `coinbaseEmptied` is now false and `newCoinbaseBalance == oldCoinbaseBalance` (the tip is 0), so `!emitCoinbase && !emitBurnt` returns nil. `K` is not re-executed — `calcFees` reads through `NewVersionedStateReader(txIndex, state.ReadSet{}, ...)`, a throwaway read set, so the coinbase is never in `K`'s `TxIn` and `J`'s write cannot invalidate it. The round-1 delete stays recorded and would wipe `J`'s credit.

**Not a regression.** `main` guards the merge the same way, one level up:

```go
tipWrites, err := txResult.calcFees(taskVer, be.versionMap, stateReader, txTask.Rules())  // main:2761
if err != nil { return nil, err }
if !tipWrites.IsEmpty() {                                                                 // main:2765
    ...MergeInto / RecordWrites / recordFeeMerge
}
```

Empty already meant "leave whatever is recorded". This PR moves that check into `recordFeeMerge` and adds a second reason for nil, so it makes the ambiguity easier to read past — fair criticism of the shape — but the stale-delete window predates it.

**Fix shape**, for whenever it lands: `calcFees` has to report which nil it means, since the caller cannot infer it (`creditedWrites` non-nil covers both branches). A three-way result — nothing-to-emit / already-recorded / new-credit — lets `recordFeeMerge` no-op on already-recorded and, on nothing-to-emit, restore the pre-merge set and release the merge product. That needs `feeMerge` to keep `prev` alongside `writes`; `prev` is the worker's `TxOut` and stays live, so holding the reference is safe.

I would rather not fold that into this PR: it changes what gets recorded, and this branch is a perf change that has not had its own node soak yet — the fee-merge direction work showed unit suites do not catch this class (715 wrong roots got through them). Happy to do it as a follow-up with an n5 soak and a regression test pinning the empty-removal transition, or here if you would rather keep it together.

[the remark this answers is not in the document]
Comment (AskAlexSharov) [reply execution/stagedsync/exec3_parallel.go]: Folded in, since leaving it out would have shipped a shape that reads as safe and is not.

`calcFees` now returns a `feeOutcome` — `feeCreditRecorded` / `feeCreditNew` / `feeCreditNone` — and `recordFeeMerge` acts on it: no-op on recorded, merge on new, retract on none. `feeMerge` keeps `base` (the worker's `TxOut`) alongside `writes`, so the retraction has something to restore.

Two things beyond the literal ask:

**The merge base.** A credit is now merged onto `base`, not onto the previous round's product. Same-shape rounds are unaffected — the new credit overwrites the old headers — but the shape can change: coinbase != burnt, zero tip, non-zero burn. Round 1 records the EIP-161 coinbase delete; round 2 emits burnt only, and `merged_N.MergeInto(tip)` keeps the delete. Rebuilding from `base` closes that alongside the transition you named, since both are "an entry the earlier credit emitted and this one does not".

**The version map.** Retracting from `blockIO` alone is not enough: the apply loop flushes the recorded set every round (`exec3_parallel.go:2877`), so the round-1 delete is already a cell later txs read. `dropStaleVersionedWrites` removes the headers the new set drops, the same header-diff the re-execution arm does at `:2738`. `checkExists=false` there because a round that returned `VersionTooEarly` never reached the flush.

`TestRecordFeeMergeRetractsVanishedCredit` pins the transition end to end — recorded set back to `TxOut`, no `SelfDestruct` cell at this txIndex, merge product released. It is red on the previous head (`Not same: ... the recorded set must fall back to the worker's own writes`).

Also took two of the non-blocking items while in here: `feeCreditRound` drives a real `blockExecutor` now instead of mirroring the round protocol — that is what made the previous fixture unable to see this bug — which collapses `TestCalcFeesRoundThroughBlockExecutor` into `TestCalcFees_SkipsRedundantReCredit`, and `BenchmarkCalcFees` gains the London row.

Package suite green. `BenchmarkCalcFees`, 200000x:

| | ns/op | allocs/op |
|---|---|---|
| pre_london/first_credit | 204 | 6 |
| pre_london/redundant_recredit | 110 | 2 |
| london/first_credit | 325 | 11 |
| london/redundant_recredit | 209 | 4 |

Backport to `release/3.6` for the merged #23131 follows once this lands.

**Review by yperbasis — CHANGES_REQUESTED** [review]
Blocking: `calcFees` still preserves a stale fee entry when only one half of the adjustment disappears.

At `exec3_parallel.go:2061`, a nil `feeEntry` reports itself as already recorded. Consider a London burn-contract transaction with a zero tip:

1. Round 1 emits an EIP-161 coinbase delete and the burnt-fee credit.
2. An earlier transaction then funds the coinbase.
3. Round 2 no longer emits the coinbase delete, while the burnt entry is unchanged.

`coinbaseEntry.recordedIn` returns true because the receiver is nil, and `burntEntry.recordedIn` returns true because that entry still matches. The combined fast path therefore returns `feeCreditRecorded`; `recordFeeMerge` does nothing, and the stale `SelfDestructPath` remains in both the recorded set and the version map. This can delete the funded coinbase and produce a wrong trie root.

The current `TestRecordFeeMerge_RetractsVanishedCredit` has no burnt entry, so every adjustment disappears and it reaches `feeCreditNone`. It does not cover this partial transition. I reproduced the issue on this head with a London variant of that test; the assertion that the coinbase `SelfDestructPath` was removed failed.

Please make the already-recorded decision verify the complete previous fee shape, including entries that are now absent, and add the partial-transition regression.

**Comment by yperbasis** [comment]
A few additions on top of the blocking review:

1. **Nothing pins the `feeCreditRecorded` early return.** Deleting `if outcome == feeCreditRecorded { return }` from `recordFeeMerge` — so every skip round runs the retraction — leaves the whole fee/finalize suite green (checked by mutation). The "already recorded" arm of `TestRecordFeeMerge_NoCreditKeepsWorkerWrites` stages no `feeMergeTemp` entry, a state `calcFees` can never pair with that outcome, and the `TestCalcFees_SkipsRedundantReCredit*` tests assert only the nil return. Worth a test that stages a real `feeCreditNew` merge, runs a Recorded round, and asserts temp survival, recorded-set identity, and the version-map entries.

2. **Release is queued one statement before the replacement is recorded.** In `recordFeeMerge`, `queueMapRelease(temp.writes)` runs before `blockIO.RecordWrites(txVersion, merged)` — the inverse of `recordWorkerWrites`' record-then-release order. In that window `blockIO.WriteSet(tx)` still returns a set whose maps the release goroutine may already be clearing. Benign while every reader sits on the apply loop; swapping the two statements removes the hazard at zero cost.

3. **Fix shape for the blocking issue: remember the emitted pair.** `feeEntry` is comparable, so `feeMerge` could store the emitted coinbase/burnt pair and `calcFees` could compare the new pair with `==` instead of probing the merged write set. `{delete, burnt}` vs `{nil, burnt}` compare unequal, so the partial transition falls into the rebuild path, which already retracts correctly. This also removes the TxNum fence in `recordedIn` (it rests on `IntraBlockState.Version()` happening to omit TxNum, enforced nowhere), frees `writeTo`/`recordedIn` from mirroring each other, lets `coinbase == burntAddr` match, and turns `dropStaleVersionedWrites` into a targeted delete of at most 4 known headers instead of an O(writes) walk on the apply loop.

4. **Stale-write deletion now lives in two loops.** `dropStaleVersionedWrites` duplicates the deletion loop in `nextResult` (walk `prev.AllHeaders()`, skip `next.Has(h)`, `versionMap.Delete`), differing only in `checkExists` and bookkeeping. One helper with a `checkExists` parameter keeps the invariant in one place; the unexplained true/false divergence otherwise reads as load-bearing.

5. **The finalize-merge record path bypasses `feeMergeTemp`.** The invariant "the recorded set is the worker's `TxOut` or `feeMergeTemp[tx].writes`" holds only because `finalizeTx` returns nil writes for regular txs. If that ever changes, a later round sees `temp.writes != prev`, uses the finalize-merged product (still embedding the old credit) as the base, and the stale-entry hazard returns while the dangling temp leaks its maps. A `dbg.AssertEnabled` check in `recordFeeMerge` is cheap. Related: `TestRecordFeeMerge_AfterReExecutionKeepsStaleTemp` stages a `temp.writes != prev` sequence that `recordWorkerWrites` makes impossible — the only real producer would be this path, so the test name points at the wrong producer.

6. Smaller points:
   - The comment on `dropStaleVersionedWrites` ("The recorded set is flushed after every round") overstates it — a `VersionTooEarly` round skips the flush, as your own `checkExists=false` rationale says. Worth aligning before someone hardens `checkExists` to true.
   - When `coinbase == burntAddr`, `burntEntry.writeTo` overwrites the coinbase entry's balance, recording pre+burn and silently dropping the tip. Pre-existing and practically unreachable on AuRa, but an alias guard emitting one combined pre+tip+burn entry would fix it and make the skip work there too.
   - The 2/4 allocs per skip round in the benchmark table are the two `vsReader.ReadAccountData` account copies in `calcFees`; an Into-style read would keep them on the stack. Follow-up material.
   - An n5 soak cannot exercise the burnt-contract arm (mainnet has no burnt contract), so the retraction fix also needs a Gnosis/Chiado soak — and the `release/3.6` backport of #23131 needs the same partial-transition fix.

File: execution/stagedsync/exec3_parallel.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (AskAlexSharov) [reply execution/stagedsync/exec3_parallel.go]: Fixed on current head. The already-recorded decision now checks the whole fee shape rather than only the entries this round emits: an address the round emits nothing for must carry no fee write of this version, so a half that stopped applying falls through to a rebuild and `recordFeeMerge` retracts it from both the recorded set and the version map.

**Comment by AskAlexSharov** [comment]
Blocking item fixed on current head, plus the three additions.

**Blocking — the partial transition.** Reproduced first: a London zero-tip tx emits the EIP-161 coinbase delete and the burnt credit, an earlier tx funds the coinbase, and round 2 emits burnt only. `coinbaseEntry` is nil and reads as recorded, `burntEntry` still matches, so the round returned `feeCreditRecorded` and the stale `SelfDestructPath` survived. `TestRecordFeeMerge_RetractsHalfOfVanishedCredit` is that scenario and was red on the old code.

The fix makes the decision check the shape, not only the entries the round emits: `shapeRecordedIn` requires an address the round emits nothing for to carry no fee write of this version (`hasFeeWrite`, discriminated by the task version's TxNum, as the delete arm already was). The round then falls through to a rebuild and `recordFeeMerge` drops the vanished write from the recorded set and the version map.

**Addition 1 — nothing pinned the `feeCreditRecorded` early return.** `TestRecordFeeMerge_RecordedRoundKeepsTheCredit` stages a real `feeCreditNew` merge, runs a Recorded round, and asserts the recorded set, the temp, the maps and the version-map entry all survive. Deleting the early return now turns it red (checked by the same mutation you used).

**Addition 2 — release ordering.** `recordFeeMerge` records the replacement before queueing the release of the set it supersedes.

**Cost.** `BenchmarkCalcFees` A/B on this machine, 6 runs: every case within noise (geomean -2.9%, p > 0.24 throughout), allocs identical — 2 for a skipped pre-London round, 4 for London.

**Backport.** As you noted this predates both PRs and is in merged #23131, so release/3.6 needs the same fix — say the word and I will open it against the twin branch.

Extra unit coverage: `TestFeeEntry_AbsentEntryMatchesOnlyAnAddressWithNoFeeWrite` pins that an absent entry matches only an address with no fee write, and that a worker's own delete (no TxNum) is not the fee shape.

**Review by yperbasis — CHANGES_REQUESTED** [review]
Blocking cleanup before merge:

The latest commit adds about 2.34 MB of unrelated local artifacts at the repository root: two screenshots and block.pprof, cpu.pprof, cpu1.pprof, fgprof.pprof, heap.pprof, and mutex.pprof. cpu1.pprof contains only a profiling error message. It also adds execution/protocol/rules/aura/contract_abi_bench_test.go, which is unrelated to the stagedsync fee-credit change.

Please remove the screenshots and profile files. If the Aura benchmarks are useful, move them to a separate PR.

I found no remaining correctness issue in the fee-credit changes. The stagedsync package tests and focused race tests pass, including against current main.

**Comment by AskAlexSharov** [comment]
Cleanup done in 3a9e1d8802 (amended onto the same commit that added them, so the branch tree no longer carries them):

- removed the two screenshots and `block.pprof`, `cpu.pprof`, `cpu1.pprof`, `fgprof.pprof`, `heap.pprof`, `mutex.pprof`
- removed `execution/protocol/rules/aura/contract_abi_bench_test.go` from this PR

The PR now touches three files, all in `execution/stagedsync`. Package suite green on the new head.

The Aura benchmarks are kept on `alex/aura_abi_bench_37`, appended to the package's existing `contract_abi_test.go` rather than a new file — happy to open that as its own PR.

**Review by yperbasis — APPROVED** [review]
Verified on 3a9e1d8802: the fee suite is green including my own variant of the repro, and the pins hold in both directions — reverting `shapeRecordedIn` to the old nil arm turns `TestRecordFeeMerge_RetractsHalfOfVanishedCredit`, my variant, and `TestFeeEntry_AbsentEntryMatchesOnlyAnAddressWithNoFeeWrite` red, and deleting the `feeCreditRecorded` early return — the mutation that survived before — now turns `TestRecordFeeMerge_RecordedRoundKeepsTheCredit` red. The release ordering matches `recordWorkerWrites` too.

The shape decision is sound: `FeeBurnt` is fixed per result, so the coinbase half is the only one that can vanish, and an absent entry now demands an address clean of this version's fee writes. The remaining cost is that `writeTo`, `recordedIn`, and `hasFeeWrite` must stay in lock-step; the new unit test pins the rejecting direction, which is the dangerous one.

Please open the `release/3.6` backport — #23131 sits there with no retraction at all, so it carries the full stale-delete bug, not only the partial transition. And since the fixed arm is the burnt-contract one, a Gnosis or Chiado soak would be evidence where the n5 soak cannot be; the n5 soak still covers the skip itself. The other items from my comment are fine as follow-ups or dropped.
